ir/json_generator.h

```diff
@@ -55,156 +57,195 @@ class JSONGenerator {
+        output_state = OBJ_START;
+        return rv;
+    }
 
-        template <class T>
-        void operator()(const T &value) const {
-            out << R"("value" : )";
-            generator.generate(value);
+    void end_object(output_state_t prev) {
+        switch (output_state) {
+            case OBJ_START:
+                out << "{}";
+                break;
+            case OBJ_MID:
+                out << std::endl << --indent << '}';
+                break;
+            case OBJ_END:
+                break;
+            default:
+                BUG("invalid json output state in end_object");
+                break;
         }
-    };
+        if ((output_state = prev) == OBJ_AFTERTAG) output_state = OBJ_MID;
+    }
 
- public:
-    indent_t indent;
+    template <typename T>
+    void emit(const T &val) {
+        switch (output_state) {
+            case VEC_MID:
+                out << ',';
+                /* fall through */
+            case VEC_START:
+                out << std::endl << indent;
+                output_state = VEC_MID;
+                break;
+            case OBJ_AFTERTAG:
+                output_state = OBJ_MID;
+                break;
+            case TOP:
+                break;
+            case OBJ_START:
+                output_state = OBJ_END;
+                break;
+            case OBJ_MID:
+            case OBJ_END:
+                BUG("invalid json output state for emit(obj)");
+        }
+        generate(val);
+        if (output_state == TOP) out << std::endl;
+    }
 
-    explicit JSONGenerator(std::ostream &out, bool dumpSourceInfo = false)
-        : out(out), dumpSourceInfo(dumpSourceInfo) {}
+    void emit_tag(std::string_view tag) {
+        switch (output_state) {
+            case OBJ_START:
+                out << '{' << std::endl << ++indent;
+                break;
+            case OBJ_MID:
+                out << ',' << std::endl << indent;
+                break;
+            case TOP:
+            case VEC_START:
+            case VEC_MID:
+            case OBJ_AFTERTAG:
+            case OBJ_END:
+                BUG("invalid json output state for emit_tag");
+        }
+        out << '\"' << cstring(tag).escapeJson() << "\" : ";
+        output_state = OBJ_AFTERTAG;
+    }
+
+    template <typename T>
+    void emit(std::string_view tag, const T &val) {
+        emit_tag(tag);
+        output_state = OBJ_MID;
+        generate(val);
+    }
 
+ private:
     template <typename T>
     void generate(const safe_vector<T> &v) {
-        out << "[";
-        if (v.size() > 0) {
-            out << std::endl << ++indent;
-            generate(v[0]);
-            for (size_t i = 1; i < v.size(); i++) {
-                out << "," << std::endl << indent;
-                generate(v[i]);
-            }
-            out << std::endl << --indent;
-        }
-        out << "]";
+        auto t = begin_vector();
+        for (auto &el : v) emit(el);
```

Comment:
maybe we can have `emit_as_vector` that would iterate over container?

Reply:
That's basically what this is.  Perhaps make a templated `generate` method that can take any iterable type (any type that has `iterator`, `begin()`, and `end()` -- C++20 concepts would be helpful here).  But there would be some ambiguity whether a type should be a vector or an object.